### ISSUE TITLE
Type instantiation is excessively deep and possibly infinite

### ISSUE THREAD

**Issue by electrovir** [blakeembrey/array-flatten#22]
Getting a ts error: `Type instantiation is excessively deep and possibly infinite. ts(2589)`

Minimal example:

```typescript
import {flatten} from 'array-flatten';

const nestedArray = [
    ['c', 'd'],
    ['e', 'f'],
];
const concatArray = ['a', 'b'].concat(flatten(nestedArray)); // error is here
console.log(concatArray);
// [ 'a', 'b', 'c', 'd', 'e', 'f' ]
```

Workaround:

```typescript
import {flatten} from 'array-flatten';

const nestedArray = [
    ['c', 'd'],
    ['e', 'f'],
];
// save off the flattened array first
const flattenedArray = flatten(nestedArray);
const concatArray = ['a', 'b'].concat(flattenedArray); // no error
console.log(concatArray);
// [ 'a', 'b', 'c', 'd', 'e', 'f' ]
```

Seems like a pretty weird edge case. I haven't looked into why it's happening. If it's not worth fixing, at least it's documented now 🤷‍♂.

**Comment by faizanu94** [comment]
@electrovir it seems to work fine (see Sandbox [here](https://codesandbox.io/s/6h62d))

**Comment by electrovir** [comment]
I can't repro anymore either 🤷‍♂️ I'm guessing a newer version of TypeScript fixed this (or one of the dependency version bumps that have happened since I hoped this bug).

Closing since can't repro.

**Comment by blakeembrey** [comment]
Thanks for the update. It might have been patched in 3.0 with https://github.com/blakeembrey/array-flatten/pull/21 which was released in November 2019 (around the same time you filed the bug, so maybe you were on 2.x?). If not, I'd also guess there's changes in TypeScript.
